### PR TITLE
capsules: add at_commands_driver

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request adds a capsule for allowing userspace to send AT commands.

### Testing Strategy

This pull request was tested by using it to control an ESP32 chip from Hail.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: capsules/src/at_commands_driver.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
+//! Userspace driver for AT commands.
+//!
+//! This driver was implemented for the ESP32 using AT commands.
```
Comment (alevy) [inline capsules/src/at_commands_driver.rs]: If it's specific to the ESP32, it ought to be named something for specific. If it's not actually specific to the ESP32.

In both cases, the comment/description should explain a bit more what the functionality is. E.g., what interface does a process get?

File: capsules/src/at_commands_driver.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                                self.uart.transmit(buffer, len);
+                            });
+                            ReturnCode::SUCCESS
+                        }
+                    })
+                }).unwrap_or_else(|err| err.into())
+            }
+            2 => {
+                self.rx_buffer.take().map_or(ReturnCode::ENOMEM, |buffer| {
+                    self.rx_in_progress.set(appid);
```
Comment (alevy) [inline capsules/src/at_commands_driver.rs]: So this is a last one wins type situation. Are AT commands transactional enough that we can just deliver the received response to the app that made the associated request?

File: capsules/src/at_commands_driver.rs (hunk 1 above)
Comment (bradjc) [reply capsules/src/at_commands_driver.rs]: Ok I updated the comment to hopefully clarify that this should be generic, and to explain a little about how it is designed to be used.

File: capsules/src/at_commands_driver.rs (hunk 2 above)
Comment (bradjc) [reply capsules/src/at_commands_driver.rs]: I think that is exactly right. Given that this is low level control of a radio (like ble-serialization) I think assuming a single app (or well coordinated apps) is a reasonable starting point.

**Comment by bradjc** [comment]
Amit raises good points that mainline Tock doesn't really need another `Driver` that doesn't have a roadmap to virtualization and proper support for multiple applications. I think this code is useful as people will probably want to use AT commands in the future (for the ESP32 and for other chips), but it's fine for it to live out-of-tree.

What would be useful is a TCP socket `Driver`...

**Pull request closed without merging**